features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp

```diff
@@ -0,0 +1,596 @@
+using namespace mbed_cellular_util;
+
+UBLOX_AT_CellularStack::UBLOX_AT_CellularStack(ATHandler &atHandler, int cid, nsapi_ip_stack_t stack_type) : AT_CellularStack(atHandler, cid, stack_type)
+{
+    // URC handlers for sockets
+    _at.set_urc_handler("+UUSORD:", callback(this, &UBLOX_AT_CellularStack::UUSORD_URC));
+    _at.set_urc_handler("+UUSORF:", callback(this, &UBLOX_AT_CellularStack::UUSORF_URC));
+    _at.set_urc_handler("+UUSOCL:", callback(this, &UBLOX_AT_CellularStack::UUSOCL_URC));
+    _at.set_urc_handler("+UUPSDD:", callback(this, &UBLOX_AT_CellularStack::UUPSDD_URC));
+	
+    pendingBytes = 0;
+}
+
+UBLOX_AT_CellularStack::~UBLOX_AT_CellularStack()
+{
+}
+
+nsapi_error_t UBLOX_AT_CellularStack::socket_listen(nsapi_socket_t handle, int backlog)
+{
+    return NSAPI_ERROR_UNSUPPORTED;
+}
+
+nsapi_error_t UBLOX_AT_CellularStack::socket_accept(void *server, void **socket, SocketAddress *addr)
+{
+    return NSAPI_ERROR_UNSUPPORTED;
+}
+
+// Callback for Socket Read URC.
+void UBLOX_AT_CellularStack::UUSORD_URC()
+{
+    int a,b;
+    CellularSocket *socket;
+
+    a =_at.read_int();
+    b =_at.read_int();
+
+    socket = find_socket(a);
+    if (socket != NULL) {
+        socket->rx_avail = true;
+        pendingBytes = b;
+        // No debug prints here as they can affect timing
+        // and cause data loss in UARTSerial
+        if (socket->_cb != NULL) {
+            socket->_cb(socket->_data);
+        }
+    }
+}
+
+// Callback for Socket Read From URC.
+void UBLOX_AT_CellularStack::UUSORF_URC()
+{
+    int a,b;
+    CellularSocket *socket;
+
+    a =_at.read_int();
+    b =_at.read_int();
+
+    socket = find_socket(a);
+    if (socket != NULL) {
+        socket->rx_avail = true;
+        pendingBytes = b;
+        // No debug prints here as they can affect timing
+        // and cause data loss in UARTSerial
+        if (socket->_cb != NULL) {
+            socket->_cb(socket->_data);
+        }
+    }
+}
+
+// Callback for Socket Close URC.
+void UBLOX_AT_CellularStack::UUSOCL_URC()
+{
+    int a;
+    CellularSocket *socket;
+
+    a =_at.read_int();
+    socket = find_socket(a);
+    clear_socket(socket);
+}
+
+// Callback for UUPSDD.
+void UBLOX_AT_CellularStack::UUPSDD_URC()
+{
+    int a;
+    CellularSocket *socket;
+
+    a =_at.read_int();
+    socket = find_socket(a);
+    clear_socket(socket);
+}
+
+
+int UBLOX_AT_CellularStack::get_max_socket_count()
+{
+    return UBLOX_MAX_SOCKET;
+}
+
+int UBLOX_AT_CellularStack::get_max_packet_size()
+{
+    return UBLOX_MAX_PACKET_SIZE;
+}
+
+bool UBLOX_AT_CellularStack::is_protocol_supported(nsapi_protocol_t protocol)
+{
+    return (protocol == NSAPI_UDP || protocol == NSAPI_TCP);
+}
+
+nsapi_error_t UBLOX_AT_CellularStack::socket_open(nsapi_socket_t *handle, nsapi_protocol_t proto)
+{
+    if (!is_protocol_supported(proto) || !handle) {
+        return NSAPI_ERROR_UNSUPPORTED;
+    }
+
+    int max_socket_count = get_max_socket_count();
+
+    if (!_socket) {
```

Comment:
I may have got this wrong, if so my apologies, but looking at the code here this seems to zero all the socket data whereas I would have expected it to open a single socket, add it to the list of open sockets in an array of CellularSocket structs, and leave all the others alone.  How does one open multiple sockets at the same time?

Reply:
Implemented virtual function of AT_CellularStack class and kept its code and added "create_socket_impl" function. This code is checking that _socket are allocated in memory or not, if not then it is creating and initializing all of them with zero.

Reply:
Ah, yes, that makes perfect sense.